### ISSUE TITLE
Docker receive app

### ISSUE THREAD

**Issue by lmars** [flynn/flynn#2763]
This is a proposal for adding a `docker-receive` app which would receive Docker images from clients and deploy them as a Flynn release.

The app would have two process types:

* `registry` - this would be the [Docker registry](https://github.com/docker/distribution/blob/master/docs/index.md) with a blobstore [storage driver](https://github.com/docker/distribution/blob/master/docs/storagedrivers.md)
* `api` - this would receive [notifications from the registry](https://github.com/docker/distribution/blob/master/docs/notifications.md) when images are pushed and create image artifacts in the controller

We would add a Flynn CLI command to wrap a call to `docker push`, subscribe to artifact events and deploy a release. We would need to come up with a way of defining the process types for the release (could be part of the CLI command, read from a local file or extracted from the Docker image).

/cc @titanous @jvatic @josephglanville thoughts?

**Comment by titanous** [comment]
This probably depends on v2 registry support, right? (#2196)

**Comment by lmars** [comment]
> This probably depends on v2 registry support, right? (#2196)

Ah yes forgot to mention, yes that is a prerequisite.

**Comment by lmars** [comment]
FYI I have a branch with an initial pass at implementing this: [docker-receive](https://github.com/flynn/flynn/compare/docker-receive).

So far I have made the following changes:

* Added a `docker-receive` app which uses the Docker registry with a [custom blobstore storage driver](https://github.com/flynn/flynn/blob/docker-receive/docker-receive/blobstore/blobstore.go) + repository middleware to [create artifacts when manifests are pushed](https://github.com/flynn/flynn/blob/docker-receive/docker-receive/main.go#L90-L106) 
* Added directory support to the blobstore (needed to implement [the storage driver `List` method](https://github.com/docker/distribution/blob/master/registry/storage/driver/storagedriver.go#L69-L71))
* Migrated pinkerton to pull from a V2 registry

There are quite a few TODOs in there, but from a high level it does what we need (see [the test](https://github.com/flynn/flynn/blob/docker-receive/test/test_docker_receive.go) :smile: ).

I'll probably clean up and make some incremental PRs soon.

**Comment by lmars** [comment]
I realised that, if we shell out to `docker push`, it will be the Docker daemon communicating with the docker-receive registry rather than the Flynn CLI, which means users not using TLS certs trusted by the Docker daemon's OS (i.e. just using the default TLS certs Flynn sets up) will have to manually configure the Docker daemon to trust those TLS certs (see [here](https://docs.docker.com/registry/insecure/#using-self-signed-certificates)).

Do we think it is acceptable to require users to configure their Docker daemon? The alternative is to require them to use TLS certs signed by a known CA.

**Comment by titanous** [comment]
That's fine. We're going to integrate Let's Encrypt in the near future, which will provide trusted certificates.

**Comment by lmars** [comment]
WRT defining process types, I decided against extracting a `Procfile` from the image because the format isn't explicitly defined, and many users expect it to be parsed by a shell (e.g. `web: bin/http -p $PORT`), but there are no guarantees that either the image or the call site of `flynn docker push` have a shell we can use to parse the line (and I don't want to write an embedded shell :wink:).

So for my initial implementation, I have decided to just grab the image config using `docker inspect` and define a single `app` process type using the relevant `CMD` / `ENTRYPOINT` / `ENV` layers, then if users need to customise the release further they can use `flynn release update`.



I think the long term approach will be to introduce a Flynnfile (#1084) and extract that from the image to define the release.
